CMakeLists.txt

```diff
@@ -1,23 +1,61 @@
-cmake_minimum_required (VERSION 2.6)
+cmake_minimum_required (VERSION 3.0...3.15)
 project (codingdojo)
 
-if (WIN32)
-
-else (WIN32) #linux
-
-	SET(Runtime_LIBRARIES pthread)
-
-endif (WIN32)
-
-include_directories(${PROJECT_SOURCE_DIR}/src/main)
-
-file(GLOB_RECURSE LIB_SOURCE src/main/*.c*)
-add_library(dojo STATIC ${LIB_SOURCE})
-
-include_directories(${PROJECT_SOURCE_DIR}/lib/gmock)
-
-add_library(gmock STATIC lib/gmock/gmock-gtest-all.cc)
-
-file(GLOB_RECURSE LIB_TEST_SOURCE src/test/*.c*)
-add_executable(dojo-test ${LIB_TEST_SOURCE} lib/gmock/gmock_main.cc)
-target_link_libraries(dojo-test dojo gmock ${Runtime_LIBRARIES})
+##################################################################
+### Enable testing and add google test helpers                 ###
+##################################################################
+enable_testing()
+include(${PROJECT_SOURCE_DIR}/cmake/AddGoogleTest.cmake)
+
+##################################################################
+### Adapt pthread for static/dynamic linking                   ###
+##################################################################
+set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
+set(THREADS_PREFER_PTHREAD_FLAG FALSE)
+find_package(Threads)
+if(TARGET Threads::Threads AND NOT BUILD_SHARED_LIBS)
+    set_target_properties(Threads::Threads PROPERTIES INTERFACE_LINK_LIBRARIES "-Wl,--whole-archive ${CMAKE_THREAD_LIBS_INIT} -Wl,--no-whole-archive")
+endif()
+
+##################################################################
+### Enable c++17 support                                       ###
+##################################################################
+add_library(project-settings INTERFACE)
+target_compile_features(project-settings INTERFACE cxx_std_17)
+
+##################################################################
+### Define sources                                             ###
+##################################################################
+add_library(${PROJECT_NAME} STATIC)
+set(PROJECT_SOURCES
+    "${CMAKE_CURRENT_SOURCE_DIR}/src/main/DummyClass.cpp"
```

Comment:
Why no globbing anymore?
This way one has to add each file manually to the build script. Does every IDE do this properly?

Reply:
Globing doesn't update the cmake cache in case new files are added, which would require a complete rebuild for each new addition. This could get cumbersome in case many files are added and deleted frequently, which can absolutely happen if you try to keep LoC low in files.

See also [this stackoverflow answer](https://stackoverflow.com/a/32412044) for an explanation.

Reply:
OK, while the intention was not to have the "ideal" template for any sort of cpp project, I'll leave it as it is now and wait for responses from users.
For comparison, I held a clean code course on c++ recently with (probably) a similar setup of CMake, and people repeatedly ran into the problem of new source files not found.